server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -928,8 +970,12 @@ private void notifyFailureOnceAllOutstandingRequestAreDone(Exception e) {
         }
     }
 
-    void createRetentionLease(final long startingSeqNo, ActionListener<RetentionLease> listener) {
-        runUnderPrimaryPermit(() -> {
+    void createRetentionLease(final long startingSeqNo, ActionListener<RetentionLease> outerListener) {
+        // NB we release the operation permit as soon as we have created the lease, but delay the outer listener until it is synced
+        final var leaseListener = new SubscribableListener<RetentionLease>();
+        final var delayedListener = outerListener.<ReplicationResponse>delegateFailure((l, ignored) -> leaseListener.addListener(l));
```

Comment:
This was the problem that https://github.com/elastic/elasticsearch/issues/95294 found: we were keeping hold of the permit until the retention lease sync completed, but that is a `ReplicationWriteOperation` so it requires another permit, and if permits are blocked in the meantime then it deadlocks and times out.

In fact this highlights (and fixes) a subtle bug in the previous implementation: if the retention lease sync happened fast enough, we could potentially complete the outer listener on this thread without releasing the primary permit. I expect this isn't a huge deal in practice since we almost certainly fork again pretty soon, but still worth noting here.